flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTaskTest.java

```diff
@@ -239,6 +244,193 @@ public void testNotMarkingEndOfInputWhenTaskCancelled () throws Exception {
 			testHarness.getOutput());
 	}
 
+	@Test
+	public void testCancellationWithSourceBlockedOnLock() throws Exception {
+		testCancellationWithSourceBlockedOnLock(false, false);
+	}
+
+	@Test
+	public void testCancellationWithSourceBlockedOnLockWithPendingMail() throws Exception {
+		testCancellationWithSourceBlockedOnLock(true, false);
+	}
+
+	@Test
+	public void testCancellationWithSourceBlockedOnLockAndThrowingOnError() throws Exception {
+		testCancellationWithSourceBlockedOnLock(false, true);
+	}
+
+	@Test
+	public void testCancellationWithSourceBlockedOnLockWithPendingMailAndThrowingOnError() throws Exception {
+		testCancellationWithSourceBlockedOnLock(true, true);
+	}
+
+	/**
+	 * Note that this test is testing also for the shared cancellation logic inside {@link StreamTask}
+	 * which, as of the time this test is being written, is not tested anywhere else
+	 * (like {@link StreamTaskTest} or {@link OneInputStreamTaskTest}).
+	 */
+	public void testCancellationWithSourceBlockedOnLock(boolean withPendingMail, boolean throwOnCancel) throws Exception {
+		final StreamTaskTestHarness<String> testHarness = new StreamTaskTestHarness<>(
+			SourceStreamTask::new,
+			BasicTypeInfo.STRING_TYPE_INFO);
+
+		CancelLockingSource.reset();
+		testHarness
+			.setupOperatorChain(
+				new OperatorID(),
+				new StreamSource<>(new CancelLockingSource(throwOnCancel)))
+			.chain(
+				new OperatorID(),
+				new TestBoundedOneInputStreamOperator("Operator1"),
+				BasicTypeInfo.STRING_TYPE_INFO.createSerializer(new ExecutionConfig()))
+			.finish();
+
+		StreamConfig streamConfig = testHarness.getStreamConfig();
+		streamConfig.setTimeCharacteristic(TimeCharacteristic.ProcessingTime);
+
+		testHarness.invoke();
+		CancelLockingSource.awaitRunning();
+		if (withPendingMail) {
+			// This pending mail should be blocked on checkpointLock acquisition, blocking the
+			// mailbox (task) thread.
+			testHarness.getTask().getMailboxExecutorFactory().createExecutor(0).execute(
+				() -> assertFalse(
+					"This should never execute before task cancelation",
+					testHarness.getTask().isRunning()),
+				"Test");
+		}
+
+		try {
+			testHarness.getTask().cancel();
+		}
+		catch (ExpectedTestException e) {
+			if (!throwOnCancel) {
+				throw e;
```

Comment:
`if !throwOnCancel then throw` sounds strange.

`throwOnCancel` -> `throwInCancel`?

Reply:
I have also replaced this with `checkState(throwInCancel);`, which also better reflects the intentions - `ExpectedTestException` shouldn't happen without `throwInCancel == true`